File: test/test_transform.py
import pytest
import molgrid

from pytest import approx

def eqQ(q1, q2):
    assert q1.R_component_1() == q2.R_component_1()
    assert q1.R_component_2() == q2.R_component_2()
    assert q1.R_component_3() == q2.R_component_3()
    assert q1.R_component_4() == q2.R_component_4()
    
def neqQ(q1, q2):
    assert q1.R_component_1() != q2.R_component_1()
    assert q1.R_component_2() != q2.R_component_2()
    assert q1.R_component_3() != q2.R_component_3()
    assert q1.R_component_4() != q2.R_component_4()
    
def tup(pt):
    '''tuplize a grid or point'''
    try:
        x = pt.x
        y = pt.y
        z = pt.z
    except AttributeError: #guess its a grid
        x = pt[0]
        y = pt[1]
        z = pt[2]
    return (x,y,z)
        
def test_random_transform():
  from molgrid import Transform
  molgrid.set_random_seed(0)
  c1 = molgrid.float3(0,0,0);
  c2 = molgrid.float3(0,0,1);

  t1 = Transform(c1, 4.0, True)
  t2 = Transform(c2, 4.0, True)

  nrt1 = Transform (c1)
  nrt2 = Transform (c2)

  t = Transform()

  molgrid.set_random_seed(0) # reset, should get same sample
  t3 = Transform(c1, 4.0, True);

  neqQ(t1.get_quaternion(),t2.get_quaternion());
  neqQ(t1.get_quaternion(),nrt1.get_quaternion());
  eqQ(t1.get_quaternion(),t3.get_quaternion());
  eqQ(nrt1.get_quaternion(),nrt2.get_quaternion());

  assert tup(t1.get_translation()) != tup(t2.get_translation())
  assert tup(t1.get_translation()) != tup(nrt1.get_translation())
  assert tup(t1.get_translation()) == tup(t3.get_translation())
  assert tup(nrt1.get_translation()) == tup(nrt2.get_translation())

  assert tup(c1) == tup(t1.get_rotation_center())
  assert tup(c2) == tup(t2.get_rotation_center())
  assert tup(c1) == tup(nrt1.get_rotation_center())

  assert tup(c1) == tup(t.get_rotation_center())
  assert tup(c1) == tup(t.get_translation())


def test_apply_transform():
    '''non-random transform'''
    from molgrid import Transform, Quaternion, float3, MGrid2f, Grid2f
    from math import sqrt
    
    q = Quaternion(sqrt(0.5),0,0,sqrt(0.5)) # //z 90
    nr = Transform(q, float3(0,1,1), float3(2,0,-3))

    #random
    r = Transform(float3(0,1,1), 10.0, True)

    coord_data = [ [0,0,0],
                   [1,0,0],
                   [0,1,0],
                   [0,0,1],
                   [-1,.5,3],
                   [1,1,1],
                   [0,1,1],
                   [.333,.75,-9] ]
  
    coords = MGrid2f(8,3)
    coords2 = MGrid2f(8,3)

    for i in range(8):
        for j in range(3):
            coords[i][j] = coord_data[i][j]

    #does nr perform as expected?
    nr.forward(coords,coords2)

    assert tup(coords2[6]) == (2,1,-2) #at center
    assert tup(coords2[2]) == (2,1,-3) #on z-axis
    assert tup(coords2[5]) == (2,2,-2)

    #make sure input unchanged
    assert tup(coords[7]) == approx((0.333,.75,-9),abs=1e-5)

    # does random work both ways
    r.forward(coords,coords2);
    for i in range(8):
        assert tup(coords[i]) != tup(coords2[i])
        
    r.backward(coords2,coords2);
    for i in range(8):
        assert tup(coords[i]) == approx(tup(coords2[i]),abs=1e-5)
        
def test_numpy_apply_transform():
    '''non-random transform'''
    from molgrid import Transform, Quaternion, float3, MGrid2f, Grid2f
    from math import sqrt
    import numpy as np
    
    q = Quaternion(sqrt(0.5),0,0,sqrt(0.5)) # //z 90
    nr = Transform(q, float3(0,1,1), float3(2,0,-3))

    #random
    r = Transform(float3(0,1,1), 10.0, True)

    coord_data = [ [0,0,0],
                   [1,0,0],
                   [0,1,0],
                   [0,0,1],
                   [-1,.5,3],
                   [1,1,1],
                   [0,1,1],
                   [.333,.75,-9] ]
  
    coords = np.array(coord_data,np.float32)
    coords2 = np.zeros((8,3),np.float32)

    #does nr perform as expected?
    nr.forward(coords,coords2)

    assert tup(coords2[6]) == (2,1,-2) #at center
    assert tup(coords2[2]) == (2,1,-3) #on z-axis
    assert tup(coords2[5]) == (2,2,-2)

    #make sure input unchanged
    assert tup(coords[7]) == approx((0.333,.75,-9),abs=1e-5)

    # does random work both ways
    r.forward(coords,coords2);
    for i in range(8):
        assert tup(coords[i]) != tup(coords2[i])
        
    r.backward(coords2,coords2);
    for i in range(8):
        assert tup(coords[i]) == approx(tup(coords2[i]),abs=1e-5)        

File: test/test_transform.cpp
/*
 * test_transform.cpp
 *
 *  Created on: Jan 10, 2019
 *      Author: dkoes
 */

#define BOOST_TEST_MODULE transform_test
#include <boost/test/unit_test.hpp>
#include <boost/test/floating_point_comparison.hpp>
#include <thrust/reduce.h>
#include <thrust/execution_policy.h>

#include "libmolgrid/libmolgrid.h"
#include "libmolgrid/transform.h"
#include "libmolgrid/managed_grid.h"
#include "test_transform.h"
using namespace libmolgrid;



BOOST_AUTO_TEST_CASE( random_transform)
{
  random_engine.seed(0);
  float3 c1 = make_float3(0,0,0);
  float3 c2 = make_float3(0,0,1);

  Transform t1(c1, 4.0, true);
  Transform t2(c2, 4.0, true);

  Transform nrt1(c1);
  Transform nrt2(c2);

  Transform t;

  random_engine.seed(0); //reset, should get same sample
  Transform t3(c1, 4.0, true);

  neqQ(t1.get_quaternion(),t2.get_quaternion());
  neqQ(t1.get_quaternion(),nrt1.get_quaternion());
  eqQ(t1.get_quaternion(),t3.get_quaternion());
  eqQ(nrt1.get_quaternion(),nrt2.get_quaternion());

  neqPt(t1.get_translation(),t2.get_translation());
  neqPt(t1.get_translation(),nrt1.get_translation());
  eqPt(t1.get_translation(),t3.get_translation());
  eqPt(nrt1.get_translation(),nrt2.get_translation());

  eqPt(c1,t1.get_rotation_center());
  eqPt(c2,t2.get_rotation_center());
  eqPt(c1,nrt1.get_rotation_center());

  eqPt(c1,t.get_rotation_center());
  eqPt(c1,t.get_translation());
}

BOOST_AUTO_TEST_CASE(apply_transform)
{
  //non-random transform
  Quaternion q(sqrt(0.5),0,0,sqrt(0.5)); //z 90
  Transform nr(q, make_float3(0,1,1), make_float3(2,0,-3));

  //random
  Transform r(make_float3(0,1,1), 10.0, true);

  float coord_data[8][3] = { {0,0,0},
                         {1,0,0},
                         {0,1,0},
                         {0,0,1},
                         {-1,.5,3},
                         {1,1,1},
                         {0,1,1},
                         {.333,.75,-9}
  };
  float buffer[8][3] = {0,};

  Grid2f coords((float*)coord_data, 8, 3);
  Grid2f coords2((float*)buffer, 8, 3);

  //does nr perform as expected?
  nr.forward(coords,coords2);
  float3 expected = make_float3(2,1,-2);
  eqPt(coords2[6],expected); //at center

  expected = make_float3(2,1,-3);
  eqPt(coords2[2],expected); //on z-axis

  expected = make_float3(2,2,-2);
  eqPt(coords2[5],expected);

  //make sure input unchanged
  expected = make_float3(0.333,.75,-9);
  eqPt(coords[7],expected);

  //does random work both ways
  r.forward(coords,coords2);
  for(unsigned i = 0; i < 8; i++) {
    neqPt(coords[i],coords2[i]);
  }

  r.backward(coords2,coords2);
  for(unsigned i = 0; i < 8; i++) {
    eqPt(coords[i],coords2[i]);
  }
}
